endpoint: create separate interface for proxy

ianvernon opened the pull request:
This is waiting on https://github.com/cilium/cilium/pull/9183 to be merged.

The `Owner` interface has served as a catch-all callback into the `Daemon` which is
managing an `Endpoint`. However, the necessary functions for proxy configuration are
specific enough to warrant their own separate interface. Create a separate interface
for the functions necessary for Endpoint<->Proxy interaction, and inject this L7 proxy
into the Endpoint upon Endpoint creation / Endpoint restoration.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9157)
<!-- Reviewable:end -->

Commit:
endpoint: create separate interface for proxy

The `Owner` interface has served as a catch-all callback into the `Daemon` which is
managing an `Endpoint`. However, the necessary functions for proxy configuration are
specific enough to warrant their own separate interface. Create a separate interface
for the functions necessary for Endpoint<->Proxy interaction, and inject this L7 proxy
into the Endpoint upon Endpoint creation / Endpoint restoration.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on daemon/state.go:
```diff
@@ -221,12 +221,13 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) (resto
 			ctmap.DeleteIfUpgradeNeeded(ep)
 		}
 
+		ep.SetProxy(d.l7Proxy)
```
This should be here https://github.com/cilium/cilium/blob/b8aabb4de02e8f2c9e73cf633aa887e73ecbf478/daemon/state.go#L163-L168

ianvernon replied:
@aanm the proxy is initiated after the call to `restoreOldEndpoints`:

restoreOldEndpoints is called here:

https://github.com/cilium/cilium/blob/8ff912d66aa8ec7d61cd6e7113ae29411963675e/daemon/daemon.go#L443-L446

proxy is initiated later:

https://github.com/cilium/cilium/blob/8ff912d66aa8ec7d61cd6e7113ae29411963675e/daemon/daemon.go#L517-L525

I could perhaps have a separate way to inject the proxy into the endpoints after L525?

ianvernon replied:
Alternatively, I can try starting the proxy before restoring endpoints. The restored endpoints aren't even exposed to the endpointmanager yet when we call `StartProxySupport`, so this should be OK.

ianvernon replied:
Done!

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -109,6 +109,8 @@ var _ notifications.RegenNotificationInfo = &Endpoint{}
 type Endpoint struct {
 	owner regeneration.Owner
 
+	proxy EndpointProxy
```
Can you put this after the `ID`? I know this sounds like a weird request but sometimes it is possible to derive the endpoint ID from stack traces, works because ID is one of the first fields of this structure.

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -78,8 +78,12 @@ func (e *Endpoint) updateNetworkPolicy(proxyWaitGroup *completion.WaitGroup) (re
 		return nil, nil
 	}
 
+	if e.proxy == nil {
```
Hmm... this is not necessarily true.. check https://github.com/cilium/cilium/issues/9100 or at least comment somewhere this is being added

ianvernon replied:
I am preserving existing behavior here: the call to `e.owner.UpdateNetworkPolicy` has now been removed. The logic in the `Owner`'s `UpdateNetworkPolicy` was the following:

```
// UpdateNetworkPolicy adds or updates a network policy in the set		
 // published to L7 proxies.		
 func (d *Daemon) UpdateNetworkPolicy(e regeneration.EndpointUpdater, policy *policy.L4Policy,		
 	proxyWaitGroup *completion.WaitGroup) (error, revert.RevertFunc) {		
 	if d.l7Proxy == nil {		
 		return fmt.Errorf("can't update network policy, proxy disabled"), nil		
 	}		
 	err, revertFunc := d.l7Proxy.UpdateNetworkPolicy(e, policy, e.GetIngressPolicyEnabledLocked(),		
 		e.GetEgressPolicyEnabledLocked(), proxyWaitGroup)		
 	return err, revert.RevertFunc(revertFunc)		
 }
```

Commit:
daemon: start proxy before restoring old endpoints

Each endpoint now has the daemon's proxy injected into it for callbacks into the
proxy during regeneration. So, on daemon bootstrap, when we restore old
endpoints, we need to inject the daemon's proxy into each restored endpoint. The
only way we can do that is if the daemon's proxy is initialized before we
restore endpoints. Do so.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: move `proxy` member after `ID`

Sometimes we can derive the endpoint ID from stack traces because ID is the
first member of Endpoint; move the proxy field lower in the Endpoint type
definition.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on daemon/daemon_test.go:
```diff
@@ -86,6 +86,9 @@ func TestMain(m *testing.M) {
 		panic("TempDir() failed.")
 	}
+
+	defer os.RemoveAll(tempRunDir)
```
This looks like a deja vu

ianvernon replied:
@aanm I updated the PR description to say that this is waiting on #9183 to be merged - the reason I did that change was for this PR to work, otherwise the proxy instantiation would fail due to the way I changed the ordering of operations for daemon bootstrap. Once #9183 is approved and merged I will drop the first commit.

ianvernon replied:
@aanm dropped this commit and rebased against master PTAL

aanm reviewed (changes requested):
All good for the last 2 commits. The first commit from this PR looks a leftover?

jrajahalme reviewed (commented):
LGTM, sorry for the late review!
